scripts/generate_matrix.py

```diff
@@ -0,0 +1,273 @@
+    )
+
+    versions: dict[tuple[int, int], Version] = {}
+
+    for release in data[::-1]:
+        version = parse(release["version"])
+
+        if version.is_prerelease and not pre_releases:
+            continue
+
+        if min_version and version < min_version:
+            continue
+        if max_version and version > max_version:
+            continue
+
+        versions[version.major, version.minor] = version
+
+    return sorted(versions.values())
+
+
+@lru_cache(maxsize=CACHE_SIZE)  # type: ignore[no-any-expr]
+def fetch_json(url: str) -> object:
+    """
+    Fetch JSON data from a URL with caching.
+
+    Args:
+        url (str): The URL to fetch.
+
+    Returns:
+        dict[str, typing.Any] | list[typing.Any]: The parsed JSON data.
+
+    Raises:
+        urllib.error.URLError: If fetching data fails.
+
+    """
+    try:
+        with urllib.request.urlopen(url) as response:  # type: ignore[no-any-expr]  # noqa: S310
+            return typing.cast("object", json.loads(response.read()))  # type: ignore[no-any-expr]
+    except urllib.error.URLError:
+        sys.exit(1)
+
+
+def get_available_package_versions(package_name: str, min_version: Version) -> dict[Version, str]:
+    """
+    Get available package versions from PyPI starting from the specified minimum version,
+    but only include the latest micro version within each minor version series,
+    along with their 'requires_python' specifiers.
+
+    Args:
+        package_name (str): The name of the package on PyPI.
+        min_version (Version): The minimum version to include.
+
+    Returns:
+        dict[Version, str]: A mapping from the latest package versions in each minor version
+                            series to their 'requires_python' specifier.
+
+    Raises:
+        RuntimeError: If no 'requires_python' is found for a package version.
+
+    """
+    data: PackageVersions = typing.cast(
+        "PackageVersions",
+        fetch_json(f"https://pypi.org/pypi/{package_name}/json"),
+    )
+
+    releases = data.get("releases", {})
+    latest_versions: dict[tuple[int, int], tuple[Version, str]] = {}
+    for version_str, release_files in releases.items():
+        version = parse(version_str)
+        if version < min_version:
+            continue
+
+            # Get 'requires_python' for this version
+        requires_python = None
+        for file_info in release_files:
+            if "requires_python" in file_info:
+                requires_python = file_info["requires_python"]
+                if requires_python:
+                    break
+
+        if requires_python is None:
+            # Skip versions without 'requires_python'
+            continue
+
+        key = (version.major, version.minor)
+        # Update to latest version within the minor version series
+        if key not in latest_versions or version > latest_versions[key][0]:
+            latest_versions[key] = (version, requires_python)
```

Comment:
```suggestion
            latest_versions[key] = version, requires_python
```

Reply:
YES! Why do many most formatters/linters/etc. love brackets so much? Are they secretly lisp-lovers? Could they be emacs users?!111cos(0)?

Reply:
🤣

Reply:
but `ruff format` doesn't add these `(...)` right 🤔?